subprojects/file-collections/src/integTest/groovy/org/gradle/api/file/FileCollectionLifecycleIntegrationTest.groovy

```diff
@@ -312,4 +313,40 @@ class FileCollectionLifecycleIntegrationTest extends AbstractIntegrationSpec imp
         outputContains("finalize failed with: Cannot finalize the value for this file collection because configuration of root project 'broken' has not completed yet.")
         output.count("value = [${file('some-file')}]") == 2
     }
+
+    @NotYetImplemented
+    def "finalizing value on read with a ConfigurableFileCollection does not lose dependency information"() {
+        buildFile """
+            abstract class Generate extends DefaultTask {
+                @OutputDirectory
+                abstract DirectoryProperty getOutputDirectory()
+                
+                @TaskAction
+                void generate() {
+                    def outputFile = outputDirectory.get().file("generated.txt").asFile
+                    outputFile.text = "generated file"
+                }
+            }
+            
+            def generateFile = tasks.register('generate', Generate) {
+                outputDirectory.convention(layout.buildDirectory)
+            }
+            
+            interface ProjectModel {
+                ConfigurableFileCollection getGeneratedFiles()
+            }
+            
+            def thing = project.extensions.create("thing", ProjectModel)
+            thing.generatedFiles.from(generateFile)
+            thing.generatedFiles.finalizeValueOnRead()
```

Comment:
If you comment out finalizeValueOnRead, then this test will pass.

The difference in behavior comes down to when we visit the file collection:
https://github.com/gradle/gradle/blob/d6c95eed025e2453bdff8b647233715d08c89f98/subprojects/file-collections/src/main/java/org/gradle/api/internal/file/collections/DefaultConfigurableFileCollection.java#L271-L279

Without `finalizeValueOnRead`, we visit the underlying file collection and see the provider to the task.

With `finalizeValueOnRead`, we visit the finalized file collection that has only `FileCollectionAdapter` and doesn't carry dependency information.

What is the expected behavior here? I think the intention is to prevent changes to the file collection, but not drop dependency information.

I can "fix" this by visiting the value before finalizing it, but that subtly changes behavior elsewhere.

Reply:
Yes, changes should be prevented and the dependency information should be preserved.